tools/validator.py

```diff
@@ -44,10 +48,23 @@ def CheckFile(self, filename):
                   artifact_definition.name, filename))
           result = False
 
-        self.defined_artifact_names.add(artifact_definition.name)
+        self._defined_artifact_names.add(artifact_definition.name)
         for source in artifact_definition.sources:
-          if isinstance(source, source_type.ArtifactSourceType):
-            self.artifact_name_references.update(source.names)
+          # TODO: change to ARTIFACT_GROUP
+          if source.type_indicator == definitions.TYPE_INDICATOR_ARTIFACT:
+            self._artifact_name_references.update(source.names)
+
+          elif source.type_indicator in (
+              definitions.TYPE_INDICATOR_WINDOWS_REGISTRY_KEY):
+            intersection = self._artifact_registry_key_paths.intersection(
+                set(source.keys))
+            if intersection:
+              duplicate_key_paths = u'\n'.join(intersection)
+              logging.warning((
```

Comment:
I think we should set result=False here. Is there a case where we want to have an exact duplicate key? Seems like we should restructure artifacts if that is the case. We should warn if we're doing fuzzy matching, but exact dups should cause an error.